sph/include/sph/table_creation.hpp

```diff
@@ -0,0 +1,163 @@
+/*! @file
+ * @brief Definition, selection and tabulation of smoothing kernels
+ *
+ * @author Sebastian Keller <sebastian.f.keller@gmail.com>
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <numeric>
+#include <vector>
+
+#include "kernels.hpp"
+
+namespace util
+{
+/*! @brief integrate a function according to simpson's rule
+ *
+ * @tparam F
+ * @param a      start of the integration interval
+ * @param b      end of the integration interval
+ * @param n      number of intervals
+ * @param func   the integrand to be integrated
+ * @return       the integral of func over [a, b]
+ */
+template<class F>
+double simpson(double a, double b, uint64_t n, F&& func)
+{
+    uint64_t numOdd  = n / 2;
+    uint64_t numEven = (numOdd >= 1) ? numOdd - 1 : 0;
+    double   h       = (b - a) / double(n);
+
+    std::vector<double> samplesOdd(numOdd);
+    std::vector<double> samplesEven(numEven);
+
+    for (uint64_t i = 0; i < numOdd; ++i)
+    {
+        uint64_t idx  = 2 * (i + 1) - 1;
+        double   x    = a + double(idx) * h;
+        samplesOdd[i] = func(x);
+    }
+    for (uint64_t i = 0; i < numEven; ++i)
+    {
+        uint64_t idx   = 2 * (i + 1);
+        double   x     = a + double(idx) * h;
+        samplesEven[i] = func(x);
+    }
+    // optional sorting for better accuracy
+    std::sort(samplesOdd.begin(), samplesOdd.end());
+    std::sort(samplesEven.begin(), samplesEven.end());
+
+    return h / 3.0 *
+           (func(a) + func(b) + 4.0 * std::accumulate(samplesOdd.begin(), samplesOdd.end(), 0.0) +
+            2.0 * std::accumulate(samplesEven.begin(), samplesEven.end(), 0.0));
+}
+} // namespace util
+
+namespace sph
+{
+
+//! @brief reference normalization constant from interpolation constants, now legacy functionality
+template<typename T>
+T sphynx_3D_k(T n)
+{
+    // b0, b1, b2 and b3 are defined in "SPHYNX: an accurate density-based SPH method for astrophysical applications",
+    // DOI: 10.1051/0004-6361/201630208
+    T b0 = 2.7012593e-2;
+    T b1 = 2.0410827e-2;
+    T b2 = 3.7451957e-3;
+    T b3 = 4.7013839e-2;
+
+    return b0 + b1 * std::sqrt(n) + b2 * n + b3 * std::sqrt(n * n * n);
+}
+
+//! @brief compute the 3D normalization constant for an arbitrary kernel
+template<class F>
+double kernel_3D_k(F&& sphKernel, double support)
+{
+    auto kernelVol3D = [sphKernel](double x) { return 4.0 * M_PI * x * x * sphKernel(x); };
+
+    uint64_t numIntervals = 2000;
```

Comment:
This number of intervals hardcoded might not be enough for some kernels if they are very steep. The integration method that you use has a way to evaluate the error. Should we add something like that to decide the numIntervals?

Reply:
It's already overkill as it is, as long as the smoothing kernel is smooth. The steepness (=first derivative) only affects  the accuracy of Riemann sums. Simpsons rule is unaffected by the steepness or curvature, because it handles anything up to cubic order exactly. That's why  the error estimation that you mentioned is proportional to the fourth derivative. That would be cumbersome to evaluate in practice. We also hardcode the number of interpolation points for the lookup table, which will become a problem much sooner than this numerical integration.